### PR TITLE
Java: Add CompositeTransport.

### PR DESCRIPTION
**Pull request by JDarDagran**
### Problem

Relates to: #2925

### Solution

Add similar to Java client.

### Checklist

- [x] You've [signed-off](https://github.com/OpenLineage/OpenLineage/blob/main/why-the-dco.md) your work
- [x] Your pull request title follows our [guidelines](https://github.com/OpenLineage/OpenLineage/blob/main/CONTRIBUTING.md#creating-pull-requests)
- [x] Your changes are accompanied by tests (_if relevant_)
- [x] Your change contains a [small diff](https://kurtisnusbaum.medium.com/stacked-diffs-keeping-phabricator-diffs-small-d9964f4dcfa6) and is self-contained
- [ ] You've updated any relevant documentation (_if relevant_)
- [ ] Your comment includes a one-liner for the changelog about the specific purpose of the change (_not required for changes to tests, docs, or CI config_)
- [ ] You've versioned the core OpenLineage model or facets according to [SchemaVer](https://docs.snowplowanalytics.com/docs/pipeline-components-and-applications/iglu/common-architecture/schemaver) (_if relevant_)
- [ ] You've added a [header](https://github.com/OpenLineage/OpenLineage/tree/main/.github/header_templates.md) to source files (_if relevant_)

----
SPDX-License-Identifier: Apache-2.0\
Copyright 2018-2024 contributors to the OpenLineage project

### CODE REVIEW

File: client/java/src/main/java/io/openlineage/client/transports/CompositeTransportBuilder.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  }
+
+  @Override
+  public Transport build(TransportConfig config) {
+    return new CompositeTransport((CompositeConfig) config);
+  }
+
+  @Override
+  public String getType() {
+    return "file";
```
Comment (pawel-big-lebowski) [inline client/java/src/main/java/io/openlineage/client/transports/CompositeTransportBuilder.java:22]: ?

File: client/java/src/main/java/io/openlineage/client/transports/CompositeTransport.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        transport.emit(datasetEvent);
+      } catch (Exception e) {
+        handleEmissionFailure(transport, e);
+      }
+    }
+  }
+
+  @Override
+  public void emit(@NonNull OpenLineage.JobEvent jobEvent) {
+    for (Transport transport : transports) {
```
Comment (pawel-big-lebowski) [inline client/java/src/main/java/io/openlineage/client/transports/CompositeTransport.java:76]: Would it make sense to use executor service in this case and emit all the transports at the same time?

Exeutor service example code: https://github.com/OpenLineage/OpenLineage/blob/main/client/java/src/main/java/io/openlineage/client/circuitBreaker/ExecutorCircuitBreaker.java#L41

**Review by pawel-big-lebowski — COMMENTED** [review]
I think it's good to have an example of the composite config in a yaml file and a test similar to: `io.openlineage.client.ConfigTest#testLoadConfigFromYaml` as yaml config for composite transport is slightly different from other transports.

Please don't hesitate to update the docs in the same PR. 

Some other comments added.

File: client/java/src/test/java/io/openlineage/client/ConfigTest.java [hunk 3]
```diff
@@ -49,6 +53,23 @@ void testLoadConfigFromYaml() throws URISyntaxException {
     assertThat(client.transport).isInstanceOf(HttpTransport.class);
   }
 
+  @SuppressWarnings({"unchecked", "PMD.AvoidAccessibilityAlteration"})
+  @Test
+  void testLoadCompositeTransportConfigFromYaml()
+      throws URISyntaxException, NoSuchFieldException, SecurityException, IllegalArgumentException,
```
Comment (arturowczarek) [inline client/java/src/test/java/io/openlineage/client/ConfigTest.java:59]: `URISyntaxException` is not thrown

File: client/java/src/test/java/io/openlineage/client/transports/CompositeTransportTest.java [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+    fakeTransportB = spy(new FakeTransportB());
+
+    // Mock configuration for the CompositeTransport
+    Map<String, Object> fakeTransportAConfig = new HashMap<>();
+    fakeTransportAConfig.put("type", "fakeA");
+    Map<String, Object> fakeTransportBConfig = new HashMap<>();
+    fakeTransportBConfig.put("type", "fakeB");
+
+    compositeConfig =
+        new CompositeConfig(List.of(fakeTransportAConfig, fakeTransportBConfig), true);
```
Comment (arturowczarek) [inline client/java/src/test/java/io/openlineage/client/transports/CompositeTransportTest.java:82]: `List.of` comes from Java 9. Does it work? In `compileTestJava` (Gradle) we have source compatibility set to java 8

Comment (arturowczarek) [reply client/java/src/test/java/io/openlineage/client/transports/CompositeTransportTest.java:82]: Apart from these little comments/doubts, looks good to me.

File: client/java/src/test/java/io/openlineage/client/ConfigTest.java (hunk 3 above)
Comment (JDarDagran) [reply client/java/src/test/java/io/openlineage/client/ConfigTest.java:59]: You're right, IDE gave me wrong hint.

Comment (JDarDagran) [reply client/java/src/test/java/io/openlineage/client/ConfigTest.java:59]: You're right, IDE gave me wrong hint.

**Review by mobuchowski — APPROVED** [review]

File: client/java/src/test/java/io/openlineage/client/transports/CompositeTransportTest.java (hunk 4 above)
Comment (mobuchowski) [reply client/java/src/test/java/io/openlineage/client/transports/CompositeTransportTest.java:82]: Arrays.asList()

Comment (JDarDagran) [reply client/java/src/test/java/io/openlineage/client/transports/CompositeTransportTest.java:82]: Changed as per @mobuchowski's suggestion.

**Review by pawel-big-lebowski — APPROVED** [review]
ready to merge once the doc is updated

File: client/java/src/main/java/io/openlineage/client/transports/CompositeConfig.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  @Getter @Setter private boolean continueOnFailure;
+
+  @JsonCreator
+  @SuppressWarnings("unchecked")
+  public CompositeConfig(
+      @JsonProperty("transports") Object transports,
+      @JsonProperty("continueOnFailure") boolean continueOnFailure) {
+
+    if (transports instanceof List) {
+      // Handle List<Map<String, Object>> case
```
Comment (pawel-big-lebowski) [inline client/java/src/main/java/io/openlineage/client/transports/CompositeConfig.java:40]: isn't it a list?

File: client/java/src/test/java/io/openlineage/client/ConfigTest.java [hunk 6]
```diff
@@ -49,6 +55,23 @@ void testLoadConfigFromYaml() throws URISyntaxException {
     assertThat(client.transport).isInstanceOf(HttpTransport.class);
   }
 
+  @SuppressWarnings({"unchecked", "PMD.AvoidAccessibilityAlteration"})
+  @ParameterizedTest
+  @ValueSource(strings = {"config/composite-array.yaml", "config/composite-map.yaml"})
```
Comment (pawel-big-lebowski) [inline client/java/src/test/java/io/openlineage/client/ConfigTest.java:60]: ❤️

File: website/docs/client/java/partials/java_transport.md [hunk 7]
```diff
@@ -558,3 +558,145 @@ OpenLineageClient client = OpenLineageClient.builder()
 
 </TabItem>
 </Tabs>
+
+## [Composite](https://github.com/OpenLineage/OpenLineage/tree/main/client/java/src/main/java/io/openlineage/client/transports/CompositeTransport.java)
+
+The `CompositeTransport` is designed to aggregate multiple transports, enabling event emission to several destinations simultaneously. This is useful when events need to be sent to more than one target, such as a logging system and an API endpoint.
```
Comment (pawel-big-lebowski) [inline website/docs/client/java/partials/java_transport.md:564]: Its current shape is not simultaneous. Isn't it?

File: client/java/src/main/java/io/openlineage/client/transports/CompositeConfig.java (hunk 5 above)
Comment (JDarDagran) [reply client/java/src/main/java/io/openlineage/client/transports/CompositeConfig.java:40]: So this is to handle both List and Map of Lists when instantianting with Jackson.
I don't like the idea of having to name each transport, you could just pass list of configurations with `name` optional. Also, currently Python version supports list only (to be changed soon when adding support for configuration through env vars).

I don't have strong opinion whether both List and List of Maps should be accepted.

**Review by pawel-big-lebowski — COMMENTED** [review]
Great docs!

File: website/docs/client/java/partials/java_transport.md (hunk 7 above)
Comment (JDarDagran) [reply website/docs/client/java/partials/java_transport.md:564]: I may reword this - I meant you can send to same event to multiple locations at once.

**Review by arturowczarek — APPROVED** [review]

**Pull request merged**